### PR TITLE
Graceful shutdown of machinepool nodes

### LINKED ISSUE #2574: AWSMachinePool graceful scale down
/kind feature
/help
/priority important-soon
/milestone v0.7.x

**Describe the solution you'd like**
When using `AWSMachinePool` and the pool is scaled down (i.e. the ASG is scaled down) i would like to ensure that the pods running on the node are evicted gracefully and that no new pods are scheduled on the node to be terminated. Currently the instance is terminated and the pods are not shutdown gracefully.

We should use a lifecycle hook on the ASG and use the notification to cordon the node and evict the running pods.

**Anything else you would like to add:**
This does not apply to `AWSManagedMachinePool` as this is handled already by AWS.

**Environment:**

- Cluster-api-provider-aws version: 0.6.6

### PR DESCRIPTION
**Pull request by akash-gautam**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
Ensures graceful shutdown of nodes in awsmachinepools (AWS ASG).
watch lifecycle transition of nodes in awsmachinepools
as soon as they enter the termination wait condition, cordon it and evict the pods
post eviction of pods complete the lifecycle hook that prevents immediate deletion of node on scale in event

**Which issue(s) this PR fixes** 
Fixes #2574 

**Release note**:

```release-note
None.
```

### CODE REVIEW

**Comment by sedefsavas** [comment]
Hi @akash-gautam, thanks for the PR. 

Few considerations here:
1- What would be the benefit of using MachinePool Machines for graceful shutdowns like CAPZ?
2- Can this implementation be extended to be used when spot instances are supported in MachinePools?
3- We already have an eventbridge implementation, but that is in experimental stage. We can graduate it to reuse here.

I think a proposal or at least an ADR would be good for this issue considering there could be multiple ways to achieve this.

cc @richardcase

**Comment by akash-gautam** [comment]
@sedefsavas 
1- What would be the benefit of using MachinePool Machines for graceful shutdowns like CAPZ? - sorry I didn't get this question can you please exaplain.
2- Can this implementation be extended to be used when spot instances are supported in MachinePools? - We can but it might be a bit tricky as the spot instances won't follow the life cycle states that on-demand instances follow in an ASG, we are creating a pre-deletion hook on the machine pool asg which will prevent the instance from getting terminated for 15 mins in case of a scale in event but for spot instance we will get only a 2 min notice (even less if the interuption action is [hibernate](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/spot-instance-termination-notices.html)) .
3- We already have an eventbridge implementation, but that is in experimental stage. We can graduate it to reuse here. - Our current implementaion is for lifecycle events of ec2 instance when is is managed my ec2 but for this we needs lifecycle events of ec2 instances which is managed by ASG. The lifecycle of ec2 instances differs based on whether they are managed by [ec2](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/ec2-instance-lifecycle.html) or [asg](https://docs.aws.amazon.com/autoscaling/ec2/userguide/AutoScalingGroupLifecycle.html). I also first tried using the [instancestate](https://github.com/kubernetes-sigs/cluster-api-provider-aws/tree/main/pkg/cloud/services/instancestate) package that we already had but for it the source of events is `aws.ec2` which doesn't capture the lifecycle states like `terminating:wait` that is why I had to create a new [package](https://github.com/akash-gautam/cluster-api-provider-aws/tree/graceful-shutdown-of-machinepool/pkg/cloud/services/asginstancestate) which is quite similar to it but caputues the events from `aws.autoscaling`.

File: exp/controllers/awsmachinepool_controller.go [hunk 1]
```diff
@@ -644,3 +666,254 @@ func (r *AWSMachinePoolReconciler) getInfraCluster(ctx context.Context, log logr
...
+func (r *AWSMachinePoolReconciler) checkForTerminatingInstancesInMachinePool(ctx context.Context, clusterScope cloud.ClusterScoper, asgsvc services.ASGInterface, asgName string) {
+
+	sqsSvs := scope.NewGlobalSQSClient(clusterScope, clusterScope)
+	queueURL, _ := r.getQueueURL(clusterScope.Name(), sqsSvs)
+	resp, err := sqsSvs.ReceiveMessage(&sqs.ReceiveMessageInput{QueueUrl: aws.String(queueURL)})
+	if err != nil {
+		clusterScope.Error(err, "non-fatal: failed to receive messages from instance state queue")
+		return
+	}
+	for _, msg := range resp.Messages {
```
[the remark this answers is not in the document]
Comment (akash-gautam) [reply exp/controllers/awsmachinepool_controller.go:679]: Yes, we can try doing something similar here as well.

**Comment by dntosas** [comment]
this is very nice feature btw 💯

**Comment by dntosas** [comment]
@sedefsavas any news on here?

**Comment by sedefsavas** [comment]
How about graduating eventbridge experimental feature and repurpose it to get event notifications for ASG scale in as well?
This will be also useful when we want to implement spot instance draining utilizing Spot Instance Termination Notice events.

@Ankitasw do you want to start a document or ADR for redesigning eventbridge flexible enough to reuse it for instance status updates + ASG scale in + spot instance termination notices? We can collaborate there and once that work is done, draining logic will just do the draining every time a relevant event is observed.

In the current eventbridge implementation, when an event is received, relevant machine reconciliation is triggered. This was enough for this use case, because machine controller can see the instance status in the reconciliation so no need to know the nature of the event. But for scale in events, probably there is no instance status change, hence events need to be processed at the controller itself.

**Comment by Ankitasw** [comment]
@sedefsavas yes I am already working on coming up with a plan on how to graduate the EventsBridge to cater status update of instances, ASG scale-in and spot instance termination notice, so as to apply those to draining logic in MachinePools and Spot Instances.

**Comment by richardcase** [comment]
> How about graduating eventbridge experimental feature and repurpose it to get event notifications for ASG scale in as well?
> This will be also useful when we want to implement spot instance draining utilizing Spot Instance Termination Notice events.

Yes i think thats a great idea @sedefsavas :+1:

**Comment by akash-gautam** [comment]
@Ankitasw If there are any tasks/issues in the graduating the EventsBridge plan that I can take up, please let me know.

**Comment by sedefsavas** [comment]
Created an issue for the eventbridge graduation https://github.com/kubernetes-sigs/cluster-api-provider-aws/issues/3414

**Comment by Ankitasw** [comment]
@akash-gautam apologies for replying late, but we have a planned a phased approach now to enable this first in EventBridge and then a separate proposal for node draining. This is the [ADR](https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/main/docs/adr/0005-graduation%20of%20EventBridge.md) for graduating EventBridge to accommodate different types of events. I would let you know once we start that feature, and see if you could make contributions on the same.

As of now closing this PR to minimize duplicate efforts, we could always revisit PR if we want to reuse some bits.
/close

**Comment by akash-gautam** [comment]
@Ankitasw Okay, thanks for the info.

**Pull request closed without merging**